codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -284,194 +316,542 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+                    }
+                    childPromise.addListener(firstFrameWriteListener);
+                    firstFrameWritten = true;
+                }
+                frame.stream(stream());
+
+                assert !childPromise.isCancellable();
+                future = ctx.write(frame);
+            } else if (msg instanceof Http2GoAwayFrame) {
+                future = ctx.write(msg);
+            } else {
+                String msgStr = msg.toString();
+                ReferenceCountUtil.release(msg);
+                throw new IllegalArgumentException(
+                        "Message must be an Http2GoAwayFrame or Http2StreamFrame: " + msgStr);
+            }
+            // Notify the child promise once complete the write
+            future.addListener(new ChannelPromiseNotifier(childPromise));
         }
 
-        Http2StreamChannel channel() {
-            return childChannel;
+        private void bytesConsumed(final int bytes) {
+            ctx.write(new DefaultHttp2WindowUpdateFrame(bytes).stream(stream()));
+        }
+
+        private static Http2StreamFrame validateStreamFrame(Object msg) {
+            if (!(msg instanceof Http2StreamFrame)) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(msg);
+                throw new IllegalArgumentException("Message must be a Http2StreamFrame: " + msgString);
+            }
+            Http2StreamFrame frame = (Http2StreamFrame) msg;
+            if (frame.stream() != null) {
+                String msgString = msg.toString();
+                ReferenceCountUtil.release(frame);
+                throw new IllegalArgumentException("Stream must not be set on the frame: " + msgString);
+            }
+            return frame;
+        }
+
+        private final class Http2ChannelUnsafe extends AbstractUnsafe {
+            @Override
+            public void connect(final SocketAddress remoteAddress,
+                                SocketAddress localAddress, final ChannelPromise promise) {
+                promise.setFailure(new UnsupportedOperationException());
+            }
+        }
+
+        /**
+         * Returns the flow-control size for DATA frames, and 0 for all other frames.
+         */
+        private static final class FlowControlledFrameSizeEstimator implements MessageSizeEstimator {
+
+            static final FlowControlledFrameSizeEstimator INSTANCE = new FlowControlledFrameSizeEstimator();
+
+            private static final class EstimatorHandle implements MessageSizeEstimator.Handle {
+
+                static final EstimatorHandle INSTANCE = new EstimatorHandle();
+
+                @Override
+                public int size(Object msg) {
+                    return msg instanceof Http2DataFrame ? ((Http2DataFrame) msg).flowControlledBytes() : 0;
+                }
+            }
+
+            @Override
+            public Handle newHandle() {
+                return EstimatorHandle.INSTANCE;
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)